File: greeks.py
"""
Black-Scholes Greeks calculator.
Computes delta, gamma, vega, theta, and charm for European options.
"""

import numpy as np
from scipy.stats import norm


def black_scholes_greeks(S, K, T, r, sigma, option_type="call", q=0.0):
    """
    Calculate Black-Scholes Greeks for a European option.

    Parameters
    ----------
    S : float – spot price of the underlying
    K : float – strike price
    T : float – time to expiration in years (must be > 0)
    r : float – risk-free interest rate (annualized, e.g. 0.05 for 5%)
    sigma : float – implied volatility (annualized, e.g. 0.20 for 20%)
    option_type : str – "call" or "put"
    q : float – continuous dividend yield (annualized)

    Returns
    -------
    dict with keys: delta, gamma, vega, theta, charm
    """
    if T <= 0 or sigma <= 0:
        return {"delta": 0.0, "gamma": 0.0, "vega": 0.0, "theta": 0.0, "charm": 0.0}

    sqrt_T = np.sqrt(T)
    d1 = (np.log(S / K) + (r - q + 0.5 * sigma ** 2) * T) / (sigma * sqrt_T)
    d2 = d1 - sigma * sqrt_T

    nd1 = norm.cdf(d1)
    nd2 = norm.cdf(d2)
    phi_d1 = norm.pdf(d1)

    # Delta
    if option_type == "call":
        delta = np.exp(-q * T) * nd1
    else:
        delta = -np.exp(-q * T) * norm.cdf(-d1)

    # Gamma (same for calls and puts)
    gamma = np.exp(-q * T) * phi_d1 / (S * sigma * sqrt_T)

    # Vega (per 1% move in vol → divide by 100)
    vega = S * np.exp(-q * T) * phi_d1 * sqrt_T / 100.0

    # Theta (per calendar day → divide by 365)
    if option_type == "call":
        theta = (
            -S * np.exp(-q * T) * phi_d1 * sigma / (2.0 * sqrt_T)
            - r * K * np.exp(-r * T) * nd2
            + q * S * np.exp(-q * T) * nd1
        ) / 365.0
    else:
        theta = (
            -S * np.exp(-q * T) * phi_d1 * sigma / (2.0 * sqrt_T)
            + r * K * np.exp(-r * T) * norm.cdf(-d2)
            - q * S * np.exp(-q * T) * norm.cdf(-d1)
        ) / 365.0

    # Charm  =  -dDelta/dT  (delta bleed per day → divide by 365)
    if option_type == "call":
        charm = -np.exp(-q * T) * (
            phi_d1 * (2.0 * (r - q) * T - d2 * sigma * sqrt_T) / (2.0 * T * sigma * sqrt_T)
            + q * nd1
        ) / 365.0
    else:
        charm = -np.exp(-q * T) * (
            phi_d1 * (2.0 * (r - q) * T - d2 * sigma * sqrt_T) / (2.0 * T * sigma * sqrt_T)
            - q * norm.cdf(-d1)
        ) / 365.0

    return {
        "delta": float(delta),
        "gamma": float(gamma),
        "vega": float(vega),
        "theta": float(theta),
        "charm": float(charm),
    }


def vectorized_greeks(S, K_arr, T, r, sigma_arr, option_types, q=0.0):
    """
    Vectorized Greeks calculation across arrays of strikes/IVs.

    Parameters
    ----------
    S : float
    K_arr : np.ndarray of strikes
    T : float
    r : float
    sigma_arr : np.ndarray of implied volatilities
    option_types : np.ndarray of "call"/"put" strings
    q : float

    Returns
    -------
    dict of np.ndarrays: delta, gamma, vega, theta, charm
    """
    K = np.asarray(K_arr, dtype=float)
    sigma = np.asarray(sigma_arr, dtype=float)
    is_call = np.array([t == "call" for t in option_types])

    valid = (T > 0) & (sigma > 0) & (K > 0)
    delta = np.zeros_like(K)
    gamma = np.zeros_like(K)
    vega = np.zeros_like(K)
    theta = np.zeros_like(K)
    charm = np.zeros_like(K)

    if not np.any(valid):
        return {"delta": delta, "gamma": gamma, "vega": vega, "theta": theta, "charm": charm}

    Kv = K[valid]
    sv = sigma[valid]
    ic = is_call[valid]
    sqrt_T = np.sqrt(T)

    d1 = (np.log(S / Kv) + (r - q + 0.5 * sv ** 2) * T) / (sv * sqrt_T)
    d2 = d1 - sv * sqrt_T
    phi_d1 = norm.pdf(d1)

    # Delta
    d = np.where(ic, np.exp(-q * T) * norm.cdf(d1), -np.exp(-q * T) * norm.cdf(-d1))
    delta[valid] = d

    # Gamma
    g = np.exp(-q * T) * phi_d1 / (S * sv * sqrt_T)
    gamma[valid] = g

    # Vega (per 1% vol)
    v = S * np.exp(-q * T) * phi_d1 * sqrt_T / 100.0
    vega[valid] = v

    # Theta (per day)
    base_theta = -S * np.exp(-q * T) * phi_d1 * sv / (2.0 * sqrt_T)
    call_theta = (base_theta - r * Kv * np.exp(-r * T) * norm.cdf(d2) + q * S * np.exp(-q * T) * norm.cdf(d1)) / 365.0
    put_theta = (base_theta + r * Kv * np.exp(-r * T) * norm.cdf(-d2) - q * S * np.exp(-q * T) * norm.cdf(-d1)) / 365.0
    theta[valid] = np.where(ic, call_theta, put_theta)

    # Charm (per day)
    charm_common = phi_d1 * (2.0 * (r - q) * T - d2 * sv * sqrt_T) / (2.0 * T * sv * sqrt_T)
    call_charm = -np.exp(-q * T) * (charm_common + q * norm.cdf(d1)) / 365.0
    put_charm = -np.exp(-q * T) * (charm_common - q * norm.cdf(-d1)) / 365.0
    charm[valid] = np.where(ic, call_charm, put_charm)

    return {
        "delta": delta,
        "gamma": gamma,
        "vega": vega,
        "theta": theta,
        "charm": charm,
    }

File: public/js/greeks.js
/**
 * Black-Scholes Greeks calculator (client-side).
 * Ports the Python greeks.py to JavaScript using the standard normal
 * distribution implemented via the rational approximation (Abramowitz & Stegun).
 */

// Standard normal PDF
function normPdf(x) {
  return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
}

// Standard normal CDF (Abramowitz & Stegun approximation, |error| < 7.5e-8)
function normCdf(x) {
  if (x === 0) return 0.5;
  const a1 = 0.254829592, a2 = -0.284496736, a3 = 1.421413741;
  const a4 = -1.453152027, a5 = 1.061405429, p = 0.3275911;
  const sign = x < 0 ? -1 : 1;
  const t = 1.0 / (1.0 + p * Math.abs(x));
  const y = 1.0 - ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-x * x / 2);
  return 0.5 * (1.0 + sign * y);
}

/**
 * Compute Black-Scholes Greeks for a single option.
 *
 * @param {number} S     - Spot price
 * @param {number} K     - Strike price
 * @param {number} T     - Time to expiry in years (> 0)
 * @param {number} r     - Risk-free rate (annualized)
 * @param {number} sigma - Implied volatility (annualized)
 * @param {string} type  - "call" or "put"
 * @param {number} [q=0] - Continuous dividend yield
 * @returns {{ delta, gamma, vega, theta, charm }}
 */
function bsGreeks(S, K, T, r, sigma, type, q = 0) {
  if (T <= 0 || sigma <= 0 || K <= 0) {
    return { delta: 0, gamma: 0, vega: 0, theta: 0, charm: 0 };
  }

  const sqrtT = Math.sqrt(T);
  const d1 = (Math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * sqrtT);
  const d2 = d1 - sigma * sqrtT;
  const phiD1 = normPdf(d1);
  const eqT = Math.exp(-q * T);
  const erT = Math.exp(-r * T);
  const isCall = type === "call";

  // Delta
  const delta = isCall
    ? eqT * normCdf(d1)
    : -eqT * normCdf(-d1);

  // Gamma (same for calls and puts)
  const gamma = eqT * phiD1 / (S * sigma * sqrtT);

  // Vega (per 1% vol move)
  const vega = S * eqT * phiD1 * sqrtT / 100.0;

  // Theta (per calendar day)
  const baseTheta = -S * eqT * phiD1 * sigma / (2.0 * sqrtT);
  const theta = isCall
    ? (baseTheta - r * K * erT * normCdf(d2) + q * S * eqT * normCdf(d1)) / 365.0
    : (baseTheta + r * K * erT * normCdf(-d2) - q * S * eqT * normCdf(-d1)) / 365.0;

  // Charm (delta bleed per day)
  const charmCommon = phiD1 * (2.0 * (r - q) * T - d2 * sigma * sqrtT) / (2.0 * T * sigma * sqrtT);
  const charm = isCall
    ? -eqT * (charmCommon + q * normCdf(d1)) / 365.0
    : -eqT * (charmCommon - q * normCdf(-d1)) / 365.0;

  return { delta, gamma, vega, theta, charm };
}

/**
 * Compute Greeks for an entire options chain array.
 * Mutates each row in-place, adding delta/gamma/vega/theta/charm fields.
 *
 * @param {Array} chain - Array of option objects with { strike, iv, type, dte }
 * @param {number} spot - Current spot price
 * @param {number} r    - Risk-free rate
 * @param {number} [q=0] - Dividend yield
 */
function computeChainGreeks(chain, spot, r, q = 0) {
  for (const row of chain) {
    const T = Math.max(row.dte, 1) / 365.0;
    const g = bsGreeks(spot, row.strike, T, r, row.iv, row.type, q);
    row.delta = g.delta;
    row.gamma = g.gamma;
    row.vega = g.vega;
    row.theta = g.theta;
    row.charm = g.charm;
  }
}
